### PR TITLE
New module pf::condition_parser for parsing conditions

### PR DESCRIPTION
**Pull request by jrouzierinverse**
## Description

Function to parse the following BNF

EXPR = OR || OR
EXPR = OR
OR   = FACT && FACT
OR   = FACT
FACT = '(' EXPR ')'
FACT = ID
ID   = /a-zA-Z0-9_/+

my $array = parse_condition_string('(a || b) && (c || d)');

Where $array would be structure

$array = [
          'AND',
          [
            'OR',
            'a',
            [
              'AND',
              'b',
              'd'
            ]
          ],
          [
            'OR',
            'c',
            'd'
          ]
        ];
## Impacts

Nothing
## NEWS file entries

New Features
++++++++++++

Enhancements
++++++++++++

Bug Fixes
+++++++++
## Delete branch after merge

NO

### CODE REVIEW

File: lib/pf/condition_parser.pm [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+}
+
+=head2 _parse_expr
+
+Handle an 'expr' expression
+
+=cut
+
+sub _parse_expr {
+# EXPR = OR || OR
```
Comment (julsemaan) [inline lib/pf/condition_parser.pm]: Fix tabulation

File: lib/pf/condition_parser.pm [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+}
+
+=head2 _parse_or
+
+Handle an 'or' expression
+
+=cut
+
+sub _parse_or {
+# OR   = FACT && FACT
```
Comment (julsemaan) [inline lib/pf/condition_parser.pm]: Tabulation

File: lib/pf/condition_parser.pm [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+}
+
+=head2 _parse_fact
+
+Handle a 'fact' expression
+
+=cut
+
+sub _parse_fact {
+# FACT = '(' EXPR ')'
```
Comment (julsemaan) [inline lib/pf/condition_parser.pm]: Tabulation

**Comment by julsemaan** [comment]
Other than the tabulation that looks good to me

**Comment by jrouzierinverse** [comment]
Please re-review

File: lib/pf/condition_parser.pm [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+                ]
+              ],
+              [
+                'OR',
+                'c',
+                'd'
+              ]
+            ];
+
+If an invalid string is passed then it will die
```
Comment (julsemaan) [inline lib/pf/condition_parser.pm]: Shouldn't it return undef instead ?

We discussed the fact that we should reduce dying and this seems a bit overkill here

Comment (jrouzierinverse) [inline lib/pf/condition_parser.pm]: It easier to report the parsing error this way.
Also returning undef will make the code a bit more complicated for something this simple.

Comment (jrouzierinverse) [inline lib/pf/condition_parser.pm]: I thought about what you said and I reworked the interface let me know what you think

**Comment by jrouzierinverse** [comment]
RE-review

**Pull request merged**